src/Interpreters/executeQuery.cpp

```diff
@@ -649,6 +649,16 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
 
         if (!async_insert)
         {
+            /// For some storage engines (e.g. ReplicatedMergeTree) we still check the value of setting
```

Comment:
@tavplubix PTAL

Reply:
Let's save shared ptr to the original context before replacing it (because we already passed that context to some places and some of them may use `ContextWeakPtr`).

Also, looks like settings in query_log will be misleading:
https://github.com/ClickHouse/ClickHouse/blob/036ddcd47baf88ab0c360efe647e01060d1ce636/src/Interpreters/executeQuery.cpp#L857

Maybe we can just update the setting without replacing context like we do here
https://github.com/ClickHouse/ClickHouse/blob/036ddcd47baf88ab0c360efe647e01060d1ce636/src/Interpreters/executeQuery.cpp#L488-L489
https://github.com/ClickHouse/ClickHouse/blob/036ddcd47baf88ab0c360efe647e01060d1ce636/src/Interpreters/executeQuery.cpp#L306
(but we still ned to do something with query_log)

Reply:
I will try to send a flag to `write` method instead because the problems you mentioned seem delicate and hard to get right.
Thanks for pointing them out